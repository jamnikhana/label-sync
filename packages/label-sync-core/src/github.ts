import * as assert from 'assert'
import Octokit from '@octokit/rest'

/**
 *
 * Github functions
 *
 */

export interface GithubRepository {
  id?: number
  owner: {
    id?: number
    login: string
  }
  name: string
  full_name: string
}

/**
 *
 * Converts Github repository name to GithubRepository
 *
 * @param name
 */
export function getRepositoryFromName(name: string): GithubRepository | null {
  try {
    const [owner, repo] = name.split('/')

    assert.ok(typeof owner === 'string')
    assert.ok(typeof repo === 'string')

    return {
      owner: { login: owner },
      name: repo,
      full_name: name,
    }
  } catch (err) {
    return null
  }
}

export interface GithubIssue {
  id: number
  number: number
  state: string
  title: string
  body: string
  labels: GithubLabel[]
}

/**
 *
 * Obtains all issues in a particular repository.
 *
 * @param github
 * @param repository
 */
export const getRepositoryIssues = paginate<GithubIssue>(
  (github, repository, page) =>
    github.issues.listForRepo({
      repo: repository.name,
      owner: repository.owner.login,
      page: page,
      per_page: 100,
    }),
  100,
)

// export interface GithubPullRequest {
//   id: number
//   number: number
//   base: { repo: GithubRepository }
//   state: string
//   title: string
//   body: string
//   labels: GithubLabel[]
// }

// /**
//  *
//  * Obtains all pull requests in a repository.
//  *
//  * @param github
//  * @param repository
//  */
// export const getRepositoryPullRequests = paginate<GithubPullRequest>(
//   (github, repository, page) =>
//     github.pulls.list({
//       repo: repository.name,
//       owner: repository.owner.login,
//       page: page,
//       per_page: 100,
//     }),
//   100,
// )

export interface GithubLabel {
  id?: number
  node_id?: string
  url?: string
  name: string
  description: string
  color: string
  default: boolean
}

/**
 *
 * Obtains all labels used in particular repository.
 *
 * @param github
 * @param repository
 */
export const getRepositoryLabels = paginate<GithubLabel>(
  (github, repository, page) =>
    github.issues.listLabelsForRepo({
      repo: repository.name,
      owner: repository.owner.login,
      page: page,
      per_page: 100,
    }),
  100,
)

/* Helper functions */

/**
 *
 * Paginates a Github function.
 *
 * @param fn
 * @param size
 */
function paginate<T>(
  fn: (
    github: Octokit,
    repository: GithubRepository,
    page: number,
  ) => Promise<Octokit.Response<T[]>>,
  size: number,
  page: number = 1,
): (github: Octokit, repository: GithubRepository) => Promise<T[]> {
  return (github: Octokit, repository: GithubRepository) =>
    fn(github, repository, page).then(async res => {
      /* Handle error case */
      if (res.status !== 200) {
        return Promise.reject()
      }

      /**
       * Return data if its size is less than treshold.
       * Paginate further otherwise.
       */
      if (res.data.length < size) {
        return res.data
      } else {
        return paginate(fn, size, page + 1)(github, repository).then(
          remainings => [...res.data, ...remainings],
        )
      }
    })
}
