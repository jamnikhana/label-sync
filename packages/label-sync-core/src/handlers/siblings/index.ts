export * from './reporter'
export * from './siblings'
export * from './sync'
