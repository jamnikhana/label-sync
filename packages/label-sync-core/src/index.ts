export { getGithubBot, BotOptions } from './tools/bot'
export { LabelConfig, Sibling, RepositoryConfig } from './types'
export {
  GithubIssue,
  GithubLabel,
  GithubRepository,
  getRepositoryFromName,
  getRepositoryIssues,
  getRepositoryLabels,
} from './github'
export { handleLabelSync } from './handlers/labels/sync'
export { handleSiblingSync } from './handlers/siblings/sync'
export { handleSync } from './tools/ci/sync'
