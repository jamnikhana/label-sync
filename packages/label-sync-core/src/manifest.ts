import lodash from 'lodash'
import * as Octokit from '@octokit/rest'
import { Sibling, RepositoryConfig, LabelConfig } from './types'
import {
  GithubLabel,
  getRepositoryLabels,
  GithubRepository,
  isLabel,
} from './github'
import { withDefault } from './utils'

export type RepositoryManifest = {
  [name: string]: LabelManifest
}

export type LabelManifest = {
  label: GithubLabel
  siblings: Sibling[]
}

/**
 *
 * Generates repository manifest from repository configuration.
 *
 * @param github
 * @param repository
 * @param config
 */
export async function getRepositoryManifest(
  github: Octokit,
  repository: GithubRepository,
  config: RepositoryConfig,
): Promise<
  | {
      status: 'ok'
      manifest: RepositoryManifest
    }
  | { status: 'err'; message: string }
> {
  /*
   * Manifest combines remote state with local configuration and warns
   * about potential errors.
   */

  const remoteLabels = await getRepositoryLabels(github, repository)
  const localLabels = getLabelsInConfiguration(config)

  let labels: GithubLabel[]

  if (config.strict) {
    labels = localLabels
  } else {
    labels = mergeLabels(remoteLabels, localLabels)
  }

  const siblings = getSiblingsInConfiguration(config)

  /* Validate configuration */

  const undefinedSiblings = siblings.filter(
    sibling => !labels.some(label => label.name === sibling),
  )

  if (undefinedSiblings.length !== 0) {
    return {
      status: 'err',
      message: `Labels ${undefinedSiblings.join(', ')} are not defined`,
    }
  }

  /* Generate manifest */

  const manifest = generateManifest(config, labels)

  return { status: 'ok', manifest: manifest }

  /* Helper functions */

  /**
   *
   * Generates repository manifest from labels and siblings.
   *
   * @param repository
   * @param labels
   */
  function generateManifest(
    repository: RepositoryConfig,
    labels: GithubLabel[],
  ): RepositoryManifest {
    const manifest = labels.reduce((acc, label) => {
      const labelConfig = repository.labels[label.name]

      switch (typeof labelConfig) {
        case 'object': {
          return {
            ...acc,
            [label.name]: {
              label: label,
              siblings: withDefault<string[]>([])(labelConfig.siblings),
            },
          }
        }
        case 'string': {
          return {
            ...acc,
            [label.name]: {
              label: label,
              siblings: [],
            },
          }
        }
        case 'undefined': {
          return {
            ...acc,
            [label.name]: {
              label: label,
              siblings: [],
            },
          }
        }
      }
    }, {})

    return manifest
  }

  /**
   *
   * Merges remote and local labels.
   *
   * @param remote
   * @param local
   */
  function mergeLabels(
    remote: GithubLabel[],
    local: GithubLabel[],
  ): GithubLabel[] {
    /* Local configuration has an upper hand over remote state. */
    return remote.reduce((acc, label) => {
      if (acc.some(isLabel(label))) {
        return acc
      } else {
        return [...acc, label]
      }
    }, local)
  }
}

/**
 *
 * Returns all siblings in a configuration.
 *
 * @param repository
 */
export function getSiblingsInConfiguration(
  repository: RepositoryConfig,
): Sibling[] {
  const siblings = Object.values(repository.labels).reduce<Sibling[]>(
    (acc, label) => {
      switch (typeof label) {
        case 'object': {
          return [...acc, ...withDefault<Sibling[]>([])(label.siblings)]
        }
        default: {
          return acc
        }
      }
    },
    [],
  )

  return lodash.uniq(siblings)
}

/**
 *
 * Returns all labels in configuration.
 *
 * @param configuration
 */
export function getLabelsInConfiguration(
  configuration: RepositoryConfig,
): GithubLabel[] {
  const labelNames = Object.keys(configuration.labels)
  const labels = labelNames.map(labelName =>
    hydrateLabel(labelName, configuration.labels[labelName]),
  )

  return labels

  /* Helper functions */

  /**
   *
   * Fills the missing pieces of a label.
   *
   * @param labelName
   * @param labelConfig
   */
  function hydrateLabel(
    labelName: string,
    labelConfig: LabelConfig,
  ): GithubLabel {
    switch (typeof labelConfig) {
      case 'string': {
        return {
          name: labelName,
          description: '',
          color: labelConfig,
          default: false,
        }
      }
      case 'object': {
        return {
          name: labelName,
          description: withDefault('')(labelConfig.description),
          color: labelConfig.color,
          default: false,
        }
      }
    }
  }
}
