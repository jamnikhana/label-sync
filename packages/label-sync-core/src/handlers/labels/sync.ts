import Octokit from '@octokit/rest'

import { RepositoryConfig } from '../../config'
import { GithubRepository, getRepositoryLabels } from '../../github'
import {
  getGithubLabelsFromRepositoryConfig,
  getLabelsDiff,
  addLabelsToRepository,
  updateLabelsInRepository,
  removeLabelsFromRepository,
} from './labels'
import { LabelSyncReport } from './reporter'

export interface LabelSyncOptions {
  dryRun: boolean
}

/**
 *
 * Handles Label Sync in a repository.
 *
 * @param client
 * @param repository
 * @param config
 * @param options
 */
export async function handleLabelSync(
  client: Octokit,
  repository: GithubRepository,
  config: RepositoryConfig,
  options: LabelSyncOptions,
): Promise<LabelSyncReport> {
  /**
   * Label Sync handler firstly loads current labels from Github,
   * and new labels from local configuration.
   *
   * After that it generates a diff and creates, updates or deletes
   * label definitions in a particular repository.
   */

  const currentLabels = await getRepositoryLabels(client, repository)
  const newLabels = getGithubLabelsFromRepositoryConfig(config)

  const diff = getLabelsDiff(currentLabels, newLabels)

  /* Sync */

  if (options.dryRun) {
    return {
      repository: repository,
      config: config,
      options: options,
      additions: diff.add,
      updates: diff.update,
      removals: diff.remove,
    }
  }

  const additions = await addLabelsToRepository(client, diff.add, repository)
  const updates = await updateLabelsInRepository(
    client,
    diff.update,
    repository,
  )
  const removals = config.strict
    ? await removeLabelsFromRepository(client, diff.remove, repository)
    : diff.remove

  return {
    repository: repository,
    config,
    options: options,
    additions,
    updates,
    removals,
  }
}
