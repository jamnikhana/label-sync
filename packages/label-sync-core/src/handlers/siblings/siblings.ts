import * as Octokit from '@octokit/rest'

import {
  GithubLabel,
  GithubRepository,
  GithubIssue,
  isLabel,
} from '../../github'
import { RepositoryManifest } from '../../manifest'
import { Sibling } from '../../types'

export type AssignSiblingsToIssueOptions = {
  dryRun: boolean
}

/**
 *
 * Assigns label siblings to an Issue.
 *
 * @param github
 * @param pr
 * @param manifest
 * @param label
 */
export async function assignSiblingsToIssue(
  github: Octokit,
  repository: GithubRepository,
  issue: GithubIssue,
  manifest: RepositoryManifest,
  options: AssignSiblingsToIssueOptions,
): Promise<GithubLabel[]> {
  /* Find all the siblings */
  const siblings = getSiblings(issue.labels, issue.labels)

  /* Only perform calculation on dryRun */
  if (options.dryRun) {
    return siblings
  }

  /* Commit changes on sync */
  const res = await github.issues.addLabels({
    repo: repository.name,
    owner: repository.owner.login,
    number: issue.number,
    labels: siblings.map(label => label.name),
  })

  /* istanbul ignore if */
  if (res.status !== 200) {
    throw new Error(JSON.stringify(res))
  }

  return siblings

  /* Helper functions */

  /**
   *
   * Get siblings recursively finds siblings of labels in a particular Github
   * issue.
   *
   * @param labels
   * @param pool
   */
  function getSiblings(
    labels: GithubLabel[],
    pool: GithubLabel[],
    path: GithubLabel[] = [],
  ): GithubLabel[] {
    /**
     * For each label in the label definitions we try to find its siblings. Each
     * sibling is then checked to be included in the calculation path. If
     * we already came accross a particular label we skip calculation to prevent
     * circular dependencies.
     *
     * Once we find the dependencies we add only the ones missing in the pool
     * to prevent duplication.
     */
    const missingSiblings = labels.reduce<GithubLabel[]>((acc, label) => {
      const combinedPool = [...acc, ...pool]

      /* Skip ciruclar dependencies */
      if (path.some(isLabel(label))) {
        return acc
      }

      /* Find manifest definition */
      const labelManifest = manifest[label.name]

      /* Filter siblings of this label */
      const newLabels = hydrateSiblings(labelManifest.siblings).filter(
        newLabel => !combinedPool.some(isLabel(newLabel)),
      )

      /* Find siblings of this label's siblings */
      const labels = getSiblings(
        hydrateSiblings(labelManifest.siblings),
        [...combinedPool, ...newLabels],
        [...path, label],
      )

      return [...acc, ...newLabels, ...labels]
    }, [])

    return missingSiblings
  }

  /**
   *
   * Hydrates siblings to Github Labels from the manifest.
   *
   * @param siblings
   */
  function hydrateSiblings(siblings: Sibling[]): GithubLabel[] {
    return siblings.map(sibling => manifest[sibling].label)
  }
}
