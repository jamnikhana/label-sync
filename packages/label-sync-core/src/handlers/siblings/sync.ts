import Octokit from '@octokit/rest'
import { SiblingSyncIssueReport, SiblingSyncReport } from './reporter'
import { assignSiblingsToIssue } from './siblings'
import {
  GithubRepository,
  getRepositoryIssues,
  GithubIssue,
} from '../../github'
import { RepositoryManifest } from '../../manifest'

export interface SiblingSyncOptions {}

/**
 *
 * Handles Sibling Sync in a repository.
 *
 * @param client
 * @param repository
 * @param config
 * @param options
 */
export async function handleSiblingSync(
  client: Octokit,
  repository: GithubRepository,
  manifest: RepositoryManifest,
  options: SiblingSyncOptions,
): Promise<SiblingSyncReport> {
  /**
   * Queries all pull requests and issues and applies siblings to existing
   * labels according to the manifest.
   */

  const issues = await getRepositoryIssues(client, repository)
  const issuesSync = await Promise.all(issues.map(handleIssue))

  return {
    repository: repository,
    manifest: manifest,
    options: options,
    issues: issuesSync,
  }

  /**
   *
   * Handles sibling sync in each issue of particular repository.
   *
   * @param issue
   */
  async function handleIssue(
    issue: GithubIssue,
  ): Promise<SiblingSyncIssueReport> {
    const siblings = await assignSiblingsToIssue(
      client,
      repository,
      issue,
      manifest,
    )

    return {
      issue: issue,
      siblings: siblings,
    }
  }
}
