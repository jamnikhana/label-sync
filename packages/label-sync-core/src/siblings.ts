import * as Octokit from '@octokit/rest'
import { Sibling, RepositoryConfig } from './config'
import {
  GithubLabel,
  getRepositoryLabels,
  GithubRepository,
  GithubIssue,
} from './github'
import { withDefault } from './utils'
import { isLabel } from './labels'

/*
 *
 * Config
 *
 */

export type RepositoryManifest = {
  [name: string]: LabelManifest
}

export type LabelManifest = {
  label: GithubLabel
  siblings: Sibling[]
}

/**
 *
 * Generates repository manifest from repository configuration.
 *
 * @param github
 * @param repository
 * @param config
 */
export async function getRepositorySiblingsManifest(
  github: Octokit,
  repository: GithubRepository,
  config: RepositoryConfig,
): Promise<
  | {
      status: 'ok'
      manifest: RepositoryManifest
    }
  | { status: 'err'; message: string }
> {
  /* Fetch repository */

  const labels = await getRepositoryLabels(github, repository)
  const siblings = getRepositorySiblings(config)

  /* Validate configuration */

  const undefinedSiblings = siblings.filter(
    sibling => !labels.some(label => label.name === sibling),
  )

  if (undefinedSiblings.length !== 0) {
    return {
      status: 'err',
      message: `${undefinedSiblings.join(', ')} are not defined`,
    }
  }

  /* Generate manifest */

  const manifest = generateManifest(config, labels)

  return { status: 'ok', manifest: manifest }

  /* Helper functions */

  /**
   *
   * Returns all siblings in a configuration.
   *
   * @param repository
   */
  function getRepositorySiblings(repository: RepositoryConfig): Sibling[] {
    const siblings = Object.values(repository.labels).reduce<Sibling[]>(
      (acc, label) => {
        switch (typeof label) {
          case 'object': {
            return [...acc, ...withDefault<Sibling[]>([])(label.siblings)]
          }
          default: {
            return acc
          }
        }
      },
      [],
    )

    return siblings
  }

  /**
   *
   * Generates repository manifest from labels and siblings.
   *
   * @param repository
   * @param labels
   */
  function generateManifest(
    repository: RepositoryConfig,
    labels: GithubLabel[],
  ): RepositoryManifest {
    const manifest = labels.reduce((acc, label) => {
      const labelConfig = repository.labels[label.name]

      switch (typeof labelConfig) {
        case 'object': {
          return {
            ...acc,
            [label.name]: {
              label: label,
              siblings: withDefault<string[]>([])(labelConfig.siblings),
            },
          }
        }
        case 'string': {
          return {
            ...acc,
            [label.name]: {
              label: label,
              siblings: [],
            },
          }
        }
        default: {
          return acc
        }
      }
    }, {})

    return manifest
  }
}

/*
 *
 * Sibling Sync
 *
 */

/**
 *
 * Assigns label siblings to an Issue.
 *
 * @param github
 * @param pr
 * @param manifest
 * @param label
 */
export async function assignSiblingsToIssue(
  github: Octokit,
  repository: GithubRepository,
  issue: GithubIssue,
  manifest: RepositoryManifest,
): Promise<
  { status: 'ok'; siblings: GithubLabel[] } | { status: 'err'; message: string }
> {
  /* Find all the siblings */
  const siblings = getSiblings(issue.labels, issue.labels)

  try {
    const res = await github.issues.addLabels({
      repo: repository.name,
      owner: repository.owner.login,
      number: issue.number,
      labels: siblings.map(label => label.name),
    })

    if (res.status === 200) {
      return { status: 'ok', siblings: siblings }
    } else {
      return { status: 'err', message: "Couldn't sync siblings." }
    }
  } catch (err) {
    return { status: 'err', message: err.message }
  }

  /* Helper functions */

  /**
   *
   * Get siblings recursively finds siblings of labels in a particular Github
   * issue.
   *
   * @param labels
   * @param pool
   */
  function getSiblings(
    labels: GithubLabel[],
    pool: GithubLabel[],
    path: GithubLabel[] = [],
  ): GithubLabel[] {
    /**
     * For each label in the label definitions we try to find its siblings. Each
     * sibling is then checked to be included in the calculation path. If
     * we already came accross a particular label we skip calculation to prevent
     * circular dependencies.
     *
     * Once we find the dependencies we add only the ones missing in the pool
     * to prevent duplication.
     */
    const missingSiblings = labels.reduce<GithubLabel[]>((acc, label) => {
      const combinedPool = [...acc, ...pool]

      /* Skip ciruclar dependencies */
      if (path.some(isLabel(label))) {
        return acc
      }

      /* Find manifest definition */
      const labelManifest = manifest[label.name]

      /* Filter siblings of this label */
      const newLabels = hydrateSiblings(labelManifest.siblings).filter(
        newLabel => !combinedPool.some(isLabel(newLabel)),
      )

      /* Find siblings of this label's siblings */
      const labels = getSiblings(
        hydrateSiblings(labelManifest.siblings),
        [...combinedPool, ...newLabels],
        [...path, label],
      )

      return [...acc, ...newLabels, ...labels]
    }, [])

    return missingSiblings
  }

  /**
   *
   * Hydrates siblings to Github Labels from the manifest.
   *
   * @param siblings
   */
  function hydrateSiblings(siblings: Sibling[]): GithubLabel[] {
    return siblings.map(sibling => manifest[sibling].label)
  }
}
