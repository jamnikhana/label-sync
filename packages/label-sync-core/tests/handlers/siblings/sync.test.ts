import { handleSiblingSync } from '../../../src/handlers/siblings/sync'
import * as fixtures from '../../__fixtures__/github'
import { getRepositoryFromName, RepositoryConfig } from '../../../src'
import { getRepositoryManifest } from '../../../src/manifest'

describe('siblings sync', () => {
  test('handles sync correctly', async () => {
    const client = fixtures.githubClient()
    const repository = getRepositoryFromName('maticzav/label-sync')!
    const config: RepositoryConfig = {
      labels: {
        'kind/bug': {
          color: '333333',
          siblings: ['bug/no-reproduction', 'kind/sibling'],
        },
        basic: 'f266f4',
        'kind/sibling': '123123',
      },
    }
    const manifest = await getRepositoryManifest(
      client as any,
      repository,
      config,
    )

    if (manifest.status === 'err') {
      fail(manifest.message)
    } else {
      const res = await handleSiblingSync(
        client as any,
        repository,
        manifest.manifest,
        { dryRun: false },
      )

      expect(client.issues.addLabels).toBeCalledTimes(2)
      expect(res).toMatchSnapshot()
    }
  })

  test('correctly performs dryrun', async () => {
    const client = fixtures.githubClient()
    const repository = getRepositoryFromName('maticzav/label-sync')!
    const config: RepositoryConfig = {
      labels: {
        'kind/bug': {
          color: '333333',
          siblings: ['bug/no-reproduction'],
        },
        basic: 'f266f4',
      },
    }
    const manifest = await getRepositoryManifest(
      client as any,
      repository,
      config,
    )

    if (manifest.status === 'err') {
      fail(manifest.message)
    } else {
      const res = await handleSiblingSync(
        client as any,
        repository,
        manifest.manifest,
        { dryRun: true },
      )

      expect(client.issues.addLabels).toBeCalledTimes(0)
      expect(res).toMatchSnapshot()
    }
  })
})
