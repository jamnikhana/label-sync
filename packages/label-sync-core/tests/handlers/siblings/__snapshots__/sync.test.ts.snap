// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`siblings sync correctly performs dryrun 1`] = `
Object {
  "issues": Array [
    Object {
      "issue": Object {
        "body": "Very serious issue.",
        "id": 1421,
        "labels": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "number": 1,
        "state": "",
        "title": "Issue Title",
      },
      "siblings": Array [],
    },
    Object {
      "issue": Object {
        "body": "Not so serious issue.",
        "id": 1123,
        "labels": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "number": 5,
        "state": "",
        "title": "Another Issue Title",
      },
      "siblings": Array [],
    },
  ],
  "manifest": Object {
    "basic": Object {
      "label": Object {
        "color": "f266f4",
        "default": false,
        "description": "",
        "name": "basic",
      },
      "siblings": Array [],
    },
    "bug/no-reproduction": Object {
      "label": Object {
        "color": "333333",
        "default": false,
        "description": "",
        "name": "bug/no-reproduction",
      },
      "siblings": Array [],
    },
    "kind/bug": Object {
      "label": Object {
        "color": "333333",
        "default": false,
        "description": "",
        "name": "kind/bug",
      },
      "siblings": Array [
        "bug/no-reproduction",
      ],
    },
  },
  "options": Object {
    "dryRun": true,
  },
  "repository": Object {
    "full_name": "maticzav/label-sync",
    "name": "label-sync",
    "owner": Object {
      "login": "maticzav",
    },
  },
}
`;

exports[`siblings sync handles sync correctly 1`] = `
Object {
  "issues": Array [
    Object {
      "issue": Object {
        "body": "Very serious issue.",
        "id": 1421,
        "labels": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "number": 1,
        "state": "",
        "title": "Issue Title",
      },
      "siblings": Array [
        Object {
          "color": "123123",
          "default": false,
          "description": "",
          "name": "kind/sibling",
        },
      ],
    },
    Object {
      "issue": Object {
        "body": "Not so serious issue.",
        "id": 1123,
        "labels": Array [
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "bug/no-reproduction",
          },
          Object {
            "color": "f266f4",
            "default": false,
            "description": "",
            "name": "basic",
          },
          Object {
            "color": "333333",
            "default": false,
            "description": "",
            "name": "kind/bug",
          },
        ],
        "number": 5,
        "state": "",
        "title": "Another Issue Title",
      },
      "siblings": Array [
        Object {
          "color": "123123",
          "default": false,
          "description": "",
          "name": "kind/sibling",
        },
      ],
    },
  ],
  "manifest": Object {
    "basic": Object {
      "label": Object {
        "color": "f266f4",
        "default": false,
        "description": "",
        "name": "basic",
      },
      "siblings": Array [],
    },
    "bug/no-reproduction": Object {
      "label": Object {
        "color": "333333",
        "default": false,
        "description": "",
        "name": "bug/no-reproduction",
      },
      "siblings": Array [],
    },
    "kind/bug": Object {
      "label": Object {
        "color": "333333",
        "default": false,
        "description": "",
        "name": "kind/bug",
      },
      "siblings": Array [
        "bug/no-reproduction",
        "kind/sibling",
      ],
    },
    "kind/sibling": Object {
      "label": Object {
        "color": "123123",
        "default": false,
        "description": "",
        "name": "kind/sibling",
      },
      "siblings": Array [],
    },
  },
  "options": Object {
    "dryRun": false,
  },
  "repository": Object {
    "full_name": "maticzav/label-sync",
    "name": "label-sync",
    "owner": Object {
      "login": "maticzav",
    },
  },
}
`;
